---
title: "R Notebook"
output: html_notebook
---


```{r}
### Figure 1A

library(reshape2)

validation_master <- read.csv("validation.tsv", sep="\t")
validation_table <- as.data.frame(table(validation_master[c(1,3)]))
validation_table2 <- dcast(validation_table, variant_id ~ method)

validation_table2[validation_table2$`RNA-Seq` > 1, 13] <- as.integer(1)
validation_table2[validation_table2$`RT-PCR` > 1, 14] <- as.integer(1)
validation_table2[,"RT-PCR"]

library(UpSetR)


svg(filename = "figures/Figure1A.svg", width = 7, height = 5)
  upset(data = validation_table2, nsets = 14, nintersects = 19, order.by = "freq", point.size = 2,	mb.ratio = c(0.5, 0.5), text.scale = 1.2,
      queries = list(list(query = intersects, params = list("MaPSy_vivo", "MaPSy_vitro"), color = "#004488", active = TRUE),
                     list(query = intersects, params = list("MFASS_SRE_dhfr", "MFASS_SRE_smn1"), color = "#004488", active = TRUE),
                     list(query = intersects, params = list("MFASS_v1", "MFASS_v2"), color = "#004488", active = TRUE),
                     list(query = intersects, params = list("MFASS_v2", "MFASS_SRE_dhfr"), color = "#004488", active = TRUE),
                     list(query = intersects, params = list("MFASS_v2", "MFASS_SRE_dhfr", "MFASS_SRE_smn1"), color = "#004488", active = TRUE),
                     list(query = intersects, params = list("MaPSy_medium5SS", "MaPSy_weak5SS", "MaPSy_strong5SS"), color = "#004488", active = TRUE),
                     list(query = intersects, params = list("MinigeneAssay", "RT-PCR"), color = "#DDAA33", active = TRUE),
                     list(query = intersects, params = list("MinigeneAssay", "MaPSy_vivo", "MaPSy_vitro"), color = "#DDAA33", active = TRUE),
                     list(query = intersects, params = list("RT-PCR", "MaPSy_vivo", "MaPSy_vitro"), color = "#DDAA33", active = TRUE),
                     list(query = intersects, params = list("MFASS_v2", "MaPSy_vivo", "MaPSy_vitro"), color = "#DDAA33", active = TRUE),
                     list(query = intersects, params = list("RNA-Seq", "MinigeneAssay"), color = "#DDAA33", active = TRUE)))
dev.off()


```


```{r}

### Figure 1B

database_master <- read.csv("outcomes.tsv", sep="\t")

ExonSkipping <- as.integer(grepl("Exon Skipping",database_master$Splicing_outcome) & (!grepl("Partial Exon Skipping",database_master$Splicing_outcome) | grepl("Partial Exon Skipping And Exon Skipping",database_master$Splicing_outcome) | grepl("Exon Skipping And Partial Exon Skipping",database_master$Splicing_outcome) | grepl("Partial Exon Skipping, Exon Skipping",database_master$Splicing_outcome) | grepl("Exon Skipping, Partial Exon Skipping",database_master$Splicing_outcome)))
Pseudoexon <-  as.integer(grepl("Pseudoexon",database_master$Splicing_outcome))
IntronRetention <-  as.integer(grepl("Intron Retention",database_master$Splicing_outcome))
PartialExonSkipping <-  as.integer(grepl("Partial Exon Skipping",database_master$Splicing_outcome))
PartialIntronInclusion <-  as.integer(grepl("Partial Intron Inclusion",database_master$Splicing_outcome))
Unreported <-  as.integer(grepl("Unreported",database_master$Splicing_outcome))

database_master[database_master$Splicing_outcome != "." & database_master$Splicing_outcome != "Unspecified" & database_master$Splicing_outcome != "1" ,]

database_outcome <- as.data.frame(cbind(ExonSkipping, Pseudoexon, IntronRetention, PartialExonSkipping, PartialIntronInclusion, Unreported))
colnames(database_outcome) <- c("Exon Skipping", "Pseudoexon Creation", "Intron Retention", "Partial Exon Skipping", "Partial Intron Inclusion", "Unreported")

svg(filename = "figures/Figure1B.svg", width = 7, height = 3)
  upset(database_outcome[1: 5],order.by = "freq", nsets = 5, nintersects = 12, point.size = 2,	mb.ratio = c(0.6, 0.4), text.scale = 1.2,
      queries = list(list(query = intersects, params = list("Exon Skipping", "Partial Exon Skipping"), color = "#BB5566", active = TRUE),
                     list(query = intersects, params = list("Exon Skipping", "Partial Intron Inclusion"), color = "#BB5566", active = TRUE),
                     list(query = intersects, params = list("Exon Skipping", "Intron Retention"), color = "#BB5566", active = TRUE),
                     list(query = intersects, params = list("Exon Skipping", "Partial Intron Inclusion", "Partial Exon Skipping"), color = "#BB5566", active = TRUE),
                     list(query = intersects, params = list("Partial Exon Skipping", "Intron Retention"), color = "#BB5566", active = TRUE),
                     list(query = intersects, params = list("Partial Exon Skipping", "Partial Intron Inclusion"), color = "#BB5566", active = TRUE),
                     list(query = intersects, params = list("Partial Intron Inclusion", "Intron Retention"), color = "#BB5566", active = TRUE)))
dev.off()

```



```{r}

# Figure2A

gene_list <- read.csv("gene_list.txt")
genes_master <- read.csv("gene_count.tsv", sep="\t")

database_genes <- as.data.frame(table(gene_list))
database_genes <- database_genes[order(database_genes$Freq, decreasing = TRUE),]

gene_plot_data <- subset(genes_master, Gene %in% database_genes[1:24,1])
gene_plot_data$Gene <- as.character(gene_plot_data$Gene)
gene_plot_data$Classification <- as.character(gene_plot_data$Category)

gene_plot_order <- as.character(database_genes[1:24,1])

gene_plot_data$Classification <- relevel(as.factor(gene_plot_data$Classification), "Low-frequency")
gene_plot_data$Classification <- relevel(as.factor(gene_plot_data$Classification), "Normal")
gene_plot_data$Classification <- relevel(as.factor(gene_plot_data$Classification), "Conflicting")

library(ggpubr)
Fig2A <- ggbarplot(gene_plot_data, "Gene", "Variants", fill="Classification", ylab = "Variant Count", xlab = "Gene", palette = c("#808080", "#3987cc", "purple", "#db3d3d"), order = gene_plot_order, ylim=c(0,500), yaxp=c(0,500,5)) + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + theme(axis.text.x = element_text(face = "italic")) + theme(legend.position = "none")


```



```{r}

### Figure 2B
regions_master <- read.csv("regions.tsv", sep="\t")

regions_master[regions_master$Region == "Deep Intronic",2] <- "Deep Intronic\n(+7 to -51)"
regions_master[regions_master$Region == "Branchpoint",2] <- "Branchpoint Region\n(-50 to -15)"
regions_master[regions_master$Region == "PPT",2] <- "PPT Region\n(-14 to -7)"
regions_master[regions_master$Region == "Acceptor Region",2] <- "Acceptor Region\n(-6 to -3)"
regions_master[regions_master$Region == "Acceptor Canonical",2] <- "Canonical Acceptor Site\n(-1 & -2)"
regions_master[regions_master$Region == "Acceptor Exonic",2] <- "Exonic Acceptor Region\n(3 exonic bases)"
regions_master[regions_master$Region == "Donor Exonic",2] <- "Exonic Donor Region\n(3 exonic bases)"
regions_master[regions_master$Region == "Donor Canonical",2] <- "Canonical Donor Site\n(+1 & +2)"
regions_master[regions_master$Region == "Donor Region",2] <- "Donor Region\n(+3 to +6)"

region_plot_data_pc <- as.data.frame(table(regions_master))
region_plot_data_pc <- subset(region_plot_data_pc, region_plot_data_pc$Region != "." & region_plot_data_pc$Classification != ".")

library(reshape2)
region_plot_data_pc <- dcast(region_plot_data_pc, Region ~ Classification)
region_plot_data_pc$Total <- rowSums(region_plot_data_pc[,2:5])
Gene_Regions <- region_plot_data_pc[,1]

region_plot_data_pc <- cbind(Gene_Regions, region_plot_data_pc[,2:5]/region_plot_data_pc[,6])
region_plot_data_pc <- melt(region_plot_data_pc)

region_plot_data_pc$variable <- relevel(as.factor(region_plot_data_pc$variable), "Low")
region_plot_data_pc$variable <- relevel(as.factor(region_plot_data_pc$variable), "Normal")
region_plot_data_pc$variable <- relevel(as.factor(region_plot_data_pc$variable), "Conflicting")

region_plot_pc <- ggbarplot(region_plot_data_pc, "Gene_Regions", "value", fill="variable", ylab = "Proportion", xlab = "Gene Region", legend.title = "Classification", palette = c("grey", "#3987cc", "purple", "#db3d3d"), order = c("Deep Intronic\n(+7 to -51)", "Branchpoint Region\n(-50 to -15)", "PPT Region\n(-14 to -7)", "Acceptor Region\n(-6 to -3)", "Canonical Acceptor Site\n(-1 & -2)", "Exonic Acceptor Region\n(3 exonic bases)", "Exonic", "Exonic Donor Region\n(3 exonic bases)", "Canonical Donor Site\n(+1 & +2)", "Donor Region\n(+3 to +6)"))

Fig2B <- region_plot_pc + theme(legend.position="none") + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + labs(fill="Legend") 


```



```{r}

### Figure 2C

regions_master <- read.csv("regions.tsv", sep="\t")

regions_master[regions_master$Region == "Deep Intronic",2] <- "Deep Intronic\n(+7 to -51)"
regions_master[regions_master$Region == "Branchpoint",2] <- "Branchpoint Region\n(-50 to -15)"
regions_master[regions_master$Region == "PPT",2] <- "PPT Region\n(-14 to -7)"
regions_master[regions_master$Region == "Acceptor Region",2] <- "Acceptor Region\n(-6 to -3)"
regions_master[regions_master$Region == "Acceptor Canonical",2] <- "Canonical Acceptor Site\n(-1 & -2)"
regions_master[regions_master$Region == "Acceptor Exonic",2] <- "Exonic Acceptor Region\n(3 exonic bases)"
regions_master[regions_master$Region == "Donor Exonic",2] <- "Exonic Donor Region\n(3 exonic bases)"
regions_master[regions_master$Region == "Donor Canonical",2] <- "Canonical Donor Site\n(+1 & +2)"
regions_master[regions_master$Region == "Donor Region",2] <- "Donor Region\n(+3 to +6)"

region_plot_data_SAV <- as.data.frame(table(regions_master[regions_master$Classification == "Splice",2]))
region_plot_data_all <- as.data.frame(table(regions_master[,2]))
region_plot_data <- merge(region_plot_data_all, region_plot_data_SAV, by="Var1")
region_plot_data <- subset(region_plot_data, region_plot_data$Var1 != ".")

region_plot <- ggbarplot(region_plot_data, "Var1", "Freq.y", fill="Var1", ylab = "Variant Count", xlab = "Gene", palette = c("grey","#347FC4", "#295A43", "#40B871", "#91CA68", "#94C6D5", "#9E8FB2", "#D6A2AD", "#DB5561", "#A61D3D"), order = c("Deep Intronic\n(+7 to -51)", "Branchpoint Region\n(-50 to -15)", "PPT Region\n(-14 to -7)", "Acceptor Region\n(-6 to -3)", "Canonical Acceptor Site\n(-1 & -2)", "Exonic Acceptor Region\n(3 exonic bases)", "Exonic", "Exonic Donor Region\n(3 exonic bases)", "Canonical Donor Site\n(+1 & +2)", "Donor Region\n(+3 to +6)"))
 
Fig2C <- region_plot + theme(legend.position="none") + theme(axis.text.x = element_text(angle = 45, hjust = 1))

```



```{r}
### Figure 2 SVG

library(ggplot2)
library(gridExtra)

Fig2A_g <- ggplotGrob(Fig2A)
Fig2B_g <- ggplotGrob(Fig2B)
Fig2C_g <- ggplotGrob(Fig2C)


Fig2 <- grid.arrange(Fig2A_g, Fig2B_g, Fig2C_g, widths = c(1,1),
                    layout_matrix = rbind(c(1,1),
                                          c(2,3)))

ggsave(filename = "figures/Figure2.svg", Fig2, width = 12)

```

